File: Diameter of Binary Tree - Leetcode 543/Diameter of Binary Tree - Leetcode 543.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

 public class Solution {
    private int largestDiameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        height(root);
        return largestDiameter;
    }

    private int height(TreeNode node) {
        if (node == null) return 0;
        
        int leftHeight = height(node.left);
        int rightHeight = height(node.right);
        int diameter = leftHeight + rightHeight;
        
        largestDiameter = Math.max(largestDiameter, diameter);
        
        return 1 + Math.max(leftHeight, rightHeight);
    }
}

// Time Complexity: O(n)
// Space Complexity: O(h) // "h" is the height of the tree

File: Diameter of Binary Tree - Leetcode 543/Diameter of Binary Tree - Leetcode 543.cpp
#include <algorithm> // For std::max
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int largestDiameter = 0;
        height(root, largestDiameter);
        return largestDiameter;
    }
    
private:
    int height(TreeNode* node, int& largestDiameter) {
        if (!node) return 0;
        
        int leftHeight = height(node->left, largestDiameter);
        int rightHeight = height(node->right, largestDiameter);
        int diameter = leftHeight + rightHeight;
        
        largestDiameter = max(largestDiameter, diameter);
        
        return 1 + max(leftHeight, rightHeight);
    }
};

// Time Complexity: O(n)
// Space Complexity: O(h) // "h" is the height of the tree
